contracts/lib.rs

```diff
@@ -955,23 +967,89 @@ pub mod prosopo {
             Ok(())
         }
 
+        /// Record a captcha result against a user, clearing out old captcha results as necessary.
+        /// A minimum of 1 captcha result will remain irrelevant of max history length or age.
+        fn record_captcha_result(&mut self, account: AccountId, result: CaptchaSolutionCommitment) -> User {
+            let mut user = self.create_new_dapp_user(account);
+            let block_timestamp = self.env().block_timestamp();
+            let max_age = if block_timestamp < self.max_user_history_age {
+                block_timestamp
+            } else {
+                self.max_user_history_age
+            };
+            let age_threshold = block_timestamp - max_age;
+            // trim the history down to max length
+            while user.history.len() > self.max_user_history_len.into() {
+                user.history.pop();
+            }
+            // trim the history down to max age (as several entrys may now be older than max age due to blocks rolling over)
+            while user.history.len() > 0 && user.history.get(user.history.len() - 1).unwrap().completed_at < age_threshold {
+                user.history.pop();
+            }
+            user.history.insert(0, result);
+            self.dapp_users.insert(account, &user);
+            user
+        }
+
+        fn get_user_history_summary(&self, account: AccountId) -> Result<UserHistorySummary, Error> {
+            let mut user = self.get_dapp_user(account)?;
+            // note that the age is based on the block timestamp, so calling this method as blocks roll over will result in different outcomes as the age threshold will change but the history will not (assuming no new results are added)
+            let block_timestamp = self.env().block_timestamp();
```

Comment:
Lines 997-1011 seem duplicated from the previous function.

Reply:
fixed in commit 274776e2